src/config_design.config

```diff
@@ -0,0 +1,126 @@
+%% Cloudant_benchmark config changes design review 
+
+%% ----------------------------------------------------------------------------
+%% CONFIGURATION TEMPLATES: (Fogbugz 62351)
+%% ----------------------------------------------------------------------------
+
+{config_templates, ["basho_bench", "cloudant_http", "http_driver"]}.
+
+%% IMPLEMENTATION
+%% Optional configuration template files to predefine parameters/preset default values
+%% Each record would contain something like (Name,Value,ValidatorFun,Description,?MODULE) 
+%%
+%% Config template files could be nested, 
+%%    would allow us to validate-on-load incoming parameters that match
+%%    warn about unknown config params to catch typos
+%%    used to auto-generate a key-value only config file for local config file use, 
+%%    would help dumping ALL config params+doc to run log, not just those provided by user
+%% 
+%% basho_bench_config:load
+%%    would pre-process terms instead of copying directly to application environmant
+%%    
+%%    if "config_templates" call then loads those config template files, 
+%%        saving parameter records into common ETS table
+%%        presetting parameter value defaults into environment
+%%
+%%    else if any templates have been defined,
+%%        checks term against known parameter records
+%%        if known then checks value with ValidatorFun before saving to environment
+%%        else warns about unknown parameters (potential typos) but accepts
+
+%% ----------------------------------------------------------------------------
+%% PER-OPERATIONS CONFIGURATION: https://github.com/cloudant/perf/issues/23 and 29
+%% ----------------------------------------------------------------------------
+%%
+%% Operations would change to allow optional per-operation keygen and valgen
+
+{operations, [
+   %% basho_bench_worker and _stats expect two values - Name, Count 
+   {get, 1},
+
+   %% Using alternate {op,[list]} format keeps 2-tuples so either format could be used
+   %% QUESTION: Do we need ONLY keygen or the OPTION of BOTH keygen and valgen ? 
+   %%    Would we ever need MORE per-op options? 
+
+   {put, [{count,1},  % Alternate count in more complex format, modify existing lookups
+          {key_generator, {int_to_str, {partitioned_sequential_int, 100000}}},
+          {value_generator, {function,http_couchdb_driver,make_json_doc,[]}}
+	 ]}
+   ]}.
+
+%% IMPLEMENTATION:
+%% basho_bench_worker:init 
+%%    turns per-op config into new Keygen/ValueGen entities
+%%    and stores into worker State as per-op-config list
+%%        per_op_config =[{put, [{keygen, KeyGen}, {valgen, ValGen}, ...]}, ... ]
+%% basho_bench_worker:worker_next_op2 
+%%    checks per_op_config, uses as run call arguments,
+%%       potentially stores onto State as current keygen/valgen
+
```

Comment:
(Russell from #cloudant-perf)

I got thinking about separating out workers and drivers so that you could organize operations by modules. For instance, instead of saying `get` here we would say `{module_foo, get}` https://github.com/cloudant/basho_bench/blob/62351-config-improvements/src/config_design.config#L39
then we could have different modules for different types of operations, say doc_crud, views, replication, etc to keep things clean. 

The tricky bit there is that currently the module is tied to the value of `driver` in the config
and similarly the new/1 function is tied to that driver as well so we would need to have per operation setup functions which is an interesting idea and might help with some of the issues for config you were talking about dschlege.

We'll still need to figure out whether we need to do the new/1 task at all. In some cases we will, for instance opening up a FD or some such but for HTTP related things we might be able to skip it entirely as we're going to use something else for setting up clusters/databases. Perhaps the best approach is to make basho_bench itself really good at executing a single benchmark and then just build in the abstractions we want in cloudant_bench.